### PR TITLE
Implement SA1604, SA1605, SA1606, SA1607

### LINKED ISSUE #124: SA1604: ElementDocumentationMustHaveSummary
http://www.stylecop.com/docs/SA1604.html

### LINKED ISSUE #125: SA1605: PartialElementDocumentationMustHaveSummary
http://www.stylecop.com/docs/SA1605.html

### LINKED ISSUE #126: SA1606: ElementDocumentationMustHaveSummaryText
http://www.stylecop.com/docs/SA1606.html

### LINKED ISSUE #127: SA1607: PartialElementDocumentationMustHaveSummaryText
http://www.stylecop.com/docs/SA1607.html

### PR DESCRIPTION
**Pull request by pdelvo**
Fixes #124 
Fixes #125 
Fixes #126 
Fixes #127

### CODE REVIEW

File: StyleCop.Analyzers/StyleCop.Analyzers/DocumentationRules/PartialElementDocumentationSummaryBase.cs [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+using StyleCop.Analyzers.Helpers;
+
+namespace StyleCop.Analyzers.DocumentationRules
+{
+    public abstract class PartialElementDocumentationSummaryBase : DiagnosticAnalyzer
+    {
+        abstract protected void HandleXmlElement(SyntaxNodeAnalysisContext context, XmlElementSyntax syntax, Location diagnosticLocation);
+        abstract protected void HandleXmlElement(SyntaxNodeAnalysisContext context, XmlElementSyntax syntax, Location[] diagnosticLocations);
```
Comment (sharwell) [inline StyleCop.Analyzers/StyleCop.Analyzers/DocumentationRules/PartialElementDocumentationSummaryBase.cs]: You can get away with just one `abstract` method here with a `params Location[] diagnosticLocations` parameter.

File: StyleCop.Analyzers/StyleCop.Analyzers/DocumentationRules/ElementDocumentationSummaryBase.cs [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+using StyleCop.Analyzers.Helpers;
+
+namespace StyleCop.Analyzers.DocumentationRules
+{
+    public abstract class ElementDocumentationSummaryBase : DiagnosticAnalyzer
+    {
+        abstract protected internal void HandleXmlElement(SyntaxNodeAnalysisContext context, XmlElementSyntax syntax, Location diagnosticLocation);
+        abstract protected internal void HandleXmlElement(SyntaxNodeAnalysisContext context, XmlElementSyntax syntax, Location[] diagnosticLocations);
```
Comment (sharwell) [inline StyleCop.Analyzers/StyleCop.Analyzers/DocumentationRules/ElementDocumentationSummaryBase.cs]: You can get away with just one `abstract` method here with a `params Location[] diagnosticLocations` parameter.

File: StyleCop.Analyzers/StyleCop.Analyzers.Test/DocumentationRules/SA1606UnitTestsy.cs [hunk 3]
```diff
@@ -0,0 +1,759 @@
+﻿namespace StyleCop.Analyzers.Test.DocumentationRules
```
Comment (sharwell) [inline StyleCop.Analyzers/StyleCop.Analyzers.Test/DocumentationRules/SA1606UnitTestsy.cs]: This file has the wrong name. :stuck_out_tongue_winking_eye:

File: StyleCop.Analyzers/StyleCop.Analyzers/DocumentationRules/ElementDocumentationSummaryBase.cs [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+            context.RegisterSyntaxNodeAction(HandleDestructorDeclaration, SyntaxKind.DestructorDeclaration);
+            context.RegisterSyntaxNodeAction(HandlePropertyDeclaration, SyntaxKind.PropertyDeclaration);
+            context.RegisterSyntaxNodeAction(HandleIndexerDeclaration, SyntaxKind.IndexerDeclaration);
+            context.RegisterSyntaxNodeAction(HandleFieldDeclaration, SyntaxKind.FieldDeclaration);
+            context.RegisterSyntaxNodeAction(HandleDelegateDeclaration, SyntaxKind.DelegateDeclaration);
+            context.RegisterSyntaxNodeAction(HandleEventDeclaration, SyntaxKind.EventDeclaration);
+            context.RegisterSyntaxNodeAction(HandleFieldDeclaration, SyntaxKind.EventFieldDeclaration);
+        }
+
+        private void HandleMemberDeclaration(SyntaxNodeAnalysisContext context)
```
Comment (sharwell) [inline StyleCop.Analyzers/StyleCop.Analyzers/DocumentationRules/ElementDocumentationSummaryBase.cs]: This should probably be renamed to `HandleTypeDeclaration`.

File: StyleCop.Analyzers/StyleCop.Analyzers/DocumentationRules/PartialElementDocumentationSummaryBase.cs [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+        /// <inheritdoc/>
+        public override void Initialize(AnalysisContext context)
+        {
+            context.RegisterSyntaxNodeAction(HandleMemberDeclaration, SyntaxKind.ClassDeclaration);
+            context.RegisterSyntaxNodeAction(HandleMemberDeclaration, SyntaxKind.StructDeclaration);
+            context.RegisterSyntaxNodeAction(HandleMemberDeclaration, SyntaxKind.InterfaceDeclaration);
+            context.RegisterSyntaxNodeAction(HandleMethodDeclaration, SyntaxKind.MethodDeclaration);
+        }
+
+        private void HandleMemberDeclaration(SyntaxNodeAnalysisContext context)
```
Comment (sharwell) [inline StyleCop.Analyzers/StyleCop.Analyzers/DocumentationRules/PartialElementDocumentationSummaryBase.cs]: This should probably be renamed to `HandleTypeDeclaration`.

**Comment by pdelvo** [comment]
I implemented your changes

**Comment by sharwell** [comment]
These are going to produce many diagnostics in code which uses `<inheritdoc/>`. Should we disable them for that case?

**Comment by pdelvo** [comment]
Jeah I think so too

**Comment by sharwell** [comment]
I'm working on the `<inheritdoc/>` support now.

**Comment by pdelvo** [comment]
Thanks! I would have implemented the inheritdoc exception myself but it was getting pretty late (3am) ;)

Looks good!

I made PartialElementDocumentationSummaryBase.HandleXmlElement internal.

File: StyleCop.Analyzers/StyleCop.Analyzers/DocumentationRules/ElementDocumentationSummaryBase.cs [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+using StyleCop.Analyzers.Helpers;
+
+namespace StyleCop.Analyzers.DocumentationRules
+{
+    public abstract class ElementDocumentationSummaryBase : DiagnosticAnalyzer
+    {
+        abstract protected internal void HandleXmlElement(SyntaxNodeAnalysisContext context, XmlNodeSyntax syntax, params Location[] diagnosticLocations);
```
Comment (sharwell) [inline StyleCop.Analyzers/StyleCop.Analyzers/DocumentationRules/ElementDocumentationSummaryBase.cs]: Why is this `protected internal`? It's not called anywhere but this class, so it seems `protected` is the better choince.

File: StyleCop.Analyzers/StyleCop.Analyzers/DocumentationRules/PartialElementDocumentationSummaryBase.cs [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+using StyleCop.Analyzers.Helpers;
+
+namespace StyleCop.Analyzers.DocumentationRules
+{
+    public abstract class PartialElementDocumentationSummaryBase : DiagnosticAnalyzer
+    {
+        abstract internal protected void HandleXmlElement(SyntaxNodeAnalysisContext context, XmlNodeSyntax syntax, params Location[] diagnosticLocations);
```
Comment (sharwell) [inline StyleCop.Analyzers/StyleCop.Analyzers/DocumentationRules/PartialElementDocumentationSummaryBase.cs]: My earlier note about `protected internal` applies here as well.

**Comment by sharwell** [comment]
> I would have implemented the inheritdoc exception myself

I have no doubt about this. I'm just trying to keep up with what y'all keep sending! :smile:

**Comment by pdelvo** [comment]
I removed the unnecessary internal modifiers

**Comment by sharwell** [comment]
You should send a new pull request enabling these diagnostics. :wink:

**Pull request merged**